NumLib/Fem/ShapeFunction/ShapeHex20-impl.h

```diff
@@ -0,0 +1,138 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2015, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+namespace
+{
+
+inline double ShapeFunctionHexHQ_Corner(const double r, const double s, const double t)
+{
+    return 0.125 * (1 + r) * (1 + s) * (1 + t) * ( r + s + t - 2.0);
+}
+
+inline double ShapeFunctionHexHQ_Middle(const double r, const double s, const double t)
+{
+    return 0.25 * (1 - r * r) * (1 + s) * (1 + t);
+}
+
+inline double dShapeFunctionHexHQ_Corner(const double r, const double s, const double t, const int ty)
+{
+    switch(ty)
+    {
+    case 0:
+        return 0.125 * (1 + s) * (1 + t) * (2.0 * r + s + t - 1.0);
+        break;
+    case 1:
+        return 0.125 * (1 + t) * (1 + r) * (2.0 * s + r + t - 1.0);
+        break;
+    case 2:
+        return 0.125 * (1 + r) * (1 + s) * (2.0 * t + s + r - 1.0);
+        break;
+    }
+    return 0.0;
+}
+
+inline double dShapeFunctionHexHQ_Middle(const double r, const double s, const double t, const int ty)
+{
+    switch(ty)
+    {
+    case 0:
+        return -0.5 * r * (1 + s) * (1 + t);
+        break;
+    case 1:
+        return 0.25 * (1 - r * r) * (1 + t);
+        break;
+    case 2:
+        return 0.25 * (1 - r * r) * (1 + s);
+        break;
+    }
+    return 0.0;
+}
+
+}
+
+namespace NumLib
+{
+
+template <class T_X, class T_N>
+void ShapeHex20::computeShapeFunction(const T_X &rst, T_N &N)
```

Comment:
yes we want to have the shape functions independent of actual data type, e.g. double[3], or Eigen::Vector or std::vector, etc.